src/rgw/rgw_common.cc

```diff
@@ -1344,6 +1345,16 @@ bool verify_bucket_permission(const DoutPrefixProvider* dpp,
     ldpp_dout(dpp, 16) << __func__ << ": policy: " << bucket_policy.get()
 		       << " resource: " << arn << dendl;
   }
+
+  // If RestrictPublicBuckets is enabled and the bucket policy allows public access,
+  // deny the request if the requester is not in the bucket owner account
+  const bool restrict_public_buckets = s->bucket_access_conf && s->bucket_access_conf->restrict_public_buckets();
+  if (restrict_public_buckets && bucket_policy && rgw::IAM::is_public(*bucket_policy) && !s->identity->is_owner_of(s->bucket_info.owner)) {
+    ldpp_dout(dpp, 10) << __func__ << ": public policies are blocked by the BlockPublicPolicy block public access setting" << dendl;
+    s->err.message = "public policies are blocked by the BlockPublicPolicy block public access setting.";
```

Comment:
is BlockPublicPolicy correct here, or should this say RestrictPublicBuckets instead?

Reply:
I'm afraid we cannot say anything due to having `struct perm_state_base * const s,`. or is there any other way of passing the message back?

Reply:
oh right, RGWUserPermHandler in rgw_data_sync.cc needs to be able to call this without a req_state. i guess you could pass an optional `std::string* message` for the callers that have one

Reply:
I was thinking of a more generic solution to that due to (https://aws.amazon.com/about-aws/whats-new/2024/08/amazon-s3-http-403-access-denied-error-messages/). Maybe worth doing it in another PR for all type of errors caused by `verify_*_permission()` and perhaps integrated to `perm_state_base`?

Reply:
i know @adamemerson has ideas to improve the client-facing diagnostics for iam policy. i certainly think it could make sense to revisit as part of that work